File: service_common.go
package spine

import (
	"fmt"
	"io"
	"log/slog"
	"net"
	"slices"

	"github.com/grandcat/zeroconf"
	"github.com/poisnoir/mad-go"
	"github.com/poisnoir/spine-go/internal/globals"
	"github.com/xtaci/kcp-go/v5"
)

// bunch of same operations in service and threaded service

func generateService[K any, V any](namespace *Namespace, name string) (*mad.Mad[K], *mad.Mad[V], *kcp.Listener, *zeroconf.Server, error) {
	logger := namespace.logger.With(
		namespace.Name(),
		"service",
		name,
		"new service",
	)

	keyEnc, err := mad.NewMad[K]()
	if err != nil {
		logger.Error("unable to create key encoder", "error", err)
		return nil, nil, nil, nil, err
	}

	valueEnc, err := mad.NewMad[V]()
	if err != nil {
		logger.Error("unable to create value encoder", "error", err)
		return nil, nil, nil, nil, err
	}

	listener, err := kcp.ListenWithOptions(":0", namespace.encryption, 10, 3)
	if err != nil {
		logger.Error("unable to create listener", "error", err)
		return nil, nil, nil, nil, err
	}

	server, err := zeroconf.Register(
		name,
		namespace.Name()+globals.ZERO_CONF_NODE_TYPE,
		globals.ZERO_CONF_DOMAIN,
		listener.Addr().(*net.UDPAddr).Port,
		[]string{"type=" + globals.ZERO_CONF_SERVICE},
		nil,
	)

	if err != nil {
		logger.Error("unable to register service to zeroconf", "error", err)
		return nil, nil, nil, nil, err
	}

	return keyEnc, valueEnc, listener, server, nil

}

func establishConnection(conn io.ReadWriteCloser, keyCode []byte, valueCode []byte, buf []byte, logger *slog.Logger) error {
	n, err := conn.Read(buf)
	if err != nil {
		return err
	}

	if !slices.Equal(keyCode, buf[:n]) {
		logger.Error("failed to establish connection")
		return fmt.Errorf("invalid key code")
	}

	_, err = conn.Write([]byte{globals.OK_STATUS_CODE})
	if err != nil {
		logger.Error("failed to establish connection")
		return err
	}

	n, err = conn.Read(buf)
	if err != nil {
		return err
	}

	if !slices.Equal(valueCode, buf[:n]) {
		logger.Error("failed to establish connection")
		return fmt.Errorf("invalid value code")
	}
	_, err = conn.Write([]byte{globals.OK_STATUS_CODE})

	return err
}

func handleCallerRequest[K any, V any](conn io.ReadWriteCloser, keySerializer *mad.Mad[K], valueSerializer *mad.Mad[V], stringSerializer *mad.Mad[string], buf []byte, processRequest func(K) serviceOutput[V], logger *slog.Logger) {

	defer conn.Close()
	err := establishConnection(conn, []byte(keySerializer.Code()), []byte(valueSerializer.Code()), buf, logger)
	if err != nil {
		return
	}

	for {
		n, err := conn.Read(buf)
		if err != nil {
			logger.Error("unable to read from connection", "error", err)
			return
		}

		switch buf[0] {
		case globals.PING_CODE:
			conn.Write([]byte{globals.PONG_CODE})

		case globals.SERVICE_REQUEST:
			var key K
			err = keySerializer.Decode(buf[1:n], &key)
			if err != nil {
				logger.Error("unable to decode key", "error", err)
				conn.Write([]byte{globals.ERROR_SERIALIZER_ERROR_CODE})
				continue
			}

			res := processRequest(key)
			if res.err != nil {
				logger.Error("handler failed", "error", err)
				errMsg := res.err.Error()
				buf[0] = globals.ERROR_SERVICE_ERROR_CODE
				stringSerializer.Encode(&errMsg, buf[1:])
				_, err = conn.Write(buf[:stringSerializer.GetRequiredSize(&errMsg)])
				if err != nil {
					logger.Error("failed to write from connection", "error", err)
					return
				}
			}

			valueSerializer.Encode(&res.data, buf)
			_, err = conn.Write(buf[:valueSerializer.GetRequiredSize(&res.data)])
			if err != nil {
				logger.Error("failed to write from connection", "error", err)
				return
			}

		default:
			logger.Error("received invalid operation code", "error", err)
			conn.Write([]byte{globals.ERROR_INVALID_OPERATION_CODE})
		}

	}
}

type serviceRequest[K any, V any] struct {
	input  K
	output chan serviceOutput[V]
}

type serviceOutput[V any] struct {
	data V
	err  error
}

File: py/service_common.py
import asyncio
import logging
import socket
from zeroconf import ServiceInfo, Zeroconf
from . import globals
from .mad import new_mad

async def generate_service(namespace, name, port=0):
    logger = namespace.logger.getChild(f"service.{name}")

    key_enc = new_mad()
    value_enc = new_mad()

    # KCP Listener setup
    # This part depends on the KCP library used.
    # We'll use a placeholder for now as we don't have the exact library.
    # In a real scenario, we'd use 'kcp.KCPServer' or similar.
    
    # Discovery registration
    zc = namespace.reg.zc
    service_type = f"{namespace.name}{globals.ZERO_CONF_TYPE}."
    full_name = f"{globals.ZERO_CONF_SERVICE_PREFIX}{name}.{service_type}"
    
    # For now, let's assume we get the local IP
    local_ip = socket.gethostbyname(socket.gethostname())
    
    info = ServiceInfo(
        type_=service_type,
        name=full_name,
        addresses=[socket.inet_aton(local_ip)],
        port=port,
        properties={'id': f'spine_service_{name}'},
        server=f"{socket.gethostname()}.local."
    )
    
    zc.register_service(info)
    
    return key_enc, value_enc, info

async def establish_connection(reader, writer, key_code, value_code, logger):
    try:
        data = await reader.read(globals.MAX_PACKET_SIZE)
        if data != key_code:
            logger.error("Failed to establish connection: invalid key code")
            return False
            
        writer.write(bytes([globals.OK_STATUS_CODE]))
        
        data = await reader.read(globals.MAX_PACKET_SIZE)
        if data != value_code:
            logger.error("Failed to establish connection: invalid value code")
            return False
            
        writer.write(bytes([globals.OK_STATUS_CODE]))
        return True
    except Exception as e:
        logger.error(f"Error establishing connection: {e}")
        return False

async def handle_caller_request(reader, writer, key_encoder, value_encoder, process_request, logger):
    # Establish connection with handshake
    if not await establish_connection(reader, writer, key_encoder.code(), value_encoder.code(), logger):
        writer.close()
        return

    try:
        while True:
            data = await reader.read(globals.MAX_PACKET_SIZE)
            if not data:
                break

            if data[0] == globals.PING_CODE:
                writer.write(bytes([globals.PONG_CODE]))
                continue

            if data[0] != globals.SERVICE_REQUEST:
                logger.error("Received invalid operation code")
                writer.write(bytes([globals.ERROR_INVALID_OPERATION_CODE]))
                continue

            # Decode key
            try:
                key = key_encoder.decode(data[1:])
            except Exception as e:
                logger.error(f"Unable to decode key: {e}")
                continue

            # Process request
            try:
                response, err = await process_request(key)
                if err:
                    logger.error(f"Handler failed: {err}")
                    writer.write(bytes([globals.ERROR_SERVICE_ERROR_CODE]))
                    continue
                
                # Encode response
                res_data = value_encoder.encode(response)
                writer.write(res_data)
            except Exception as e:
                logger.error(f"Process request failed: {e}")
                writer.write(bytes([globals.ERROR_HANDLER_INTERNAL_ERROR_CODE]))
                
    except asyncio.CancelledError:
        pass
    except Exception as e:
        logger.error(f"Connection error: {e}")
    finally:
        writer.close()
